### PR TITLE
[examples][MLIRVector] improve vector dialect examples for create and cast

### PR DESCRIPTION
**Pull request by Origami404**
add or improve examples for operations related to create normal or mask vector and cast:

- splat
- broadcast
- create_mask
- constant_mask
- bitcast
- shape_cast
- type_cast

This PR is a small and finished part of PR #90

### CODE REVIEW

File: examples/MLIRVector/vector-bitcast.mlir [hunk 1]
```diff
@@ -1,16 +1,60 @@
...
   vector.print %v3 : vector<6xi32>
 
+  // bitcast could only be used between vector types with
+  // same total length in byte, like 8xi32 <-> 4xf64
+
+  // error: 'vector.bitcast' op source/result bitwidth of the minor 1-D vectors must be equal
+  // %v4 = vector.bitcast %v0 : vector<6xi32> to vector<4xi64> 
+
+
+  // Because scalable vector is platform specify, vector dialect could not
```
Comment (meshtag) [inline examples/MLIRVector/vector-bitcast.mlir:33]: Typo: specify -> specific

File: examples/MLIRVector/vector-bitcast.mlir [hunk 2]
```diff
@@ -1,16 +1,60 @@
...
+  // lowering/translating them well, so we just assume that we have one:
```
Comment (meshtag) [inline examples/MLIRVector/vector-bitcast.mlir:34]: Typo: lowering/translating -> lower/translate

File: examples/MLIRVector/vector-bitcast.mlir [hunk 3]
```diff
@@ -1,16 +1,60 @@
...
+  // %v7 = vector.bitcast %v5 : vector<[4]xi32> to vector<[8]xi16>
+  // vector.print %v7 : vector<[8]xi16>
+
+  // bitcast operations of scalable dimensions must respect the bitwidth
+  // restriction as same with fix length dimensions 
+
+  // error: 'vector.bitcast' op source/result bitwidth of the minor 1-D vectors must be equal
+  // %v8 = vector.bitcast %v5 : vector<[4]xi32> to vector<[3]xi64>
+
+  // bitcast operations of scalable dimensions should ALWAYS meet the
```
Comment (meshtag) [inline examples/MLIRVector/vector-bitcast.mlir:52]: Is there a need for this example to be specified separately?

File: examples/MLIRVector/vector-broadcast.mlir [hunk 4]
```diff
@@ -17,6 +19,117 @@ func.func @main() -> i32 {
...
+
   %ret = arith.constant 0 : i32
   return %ret : i32
 }
+
+func.func @broadcast_scalar_to_vector(%src: f32) -> vector<3x1xf32> {
+  // broadcast a scalar to vector
+  %result = vector.broadcast %src : f32 to vector<3x1xf32>
+
+  // equals to splat it, or, "fill" it
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:39]: This is not completely clear. Can you please rephrase it.

File: examples/MLIRVector/vector-broadcast.mlir [hunk 5]
```diff
@@ -17,6 +19,117 @@ func.func @main() -> i32 {
...
+
+  // %w == %result
+  vector.print %result : vector<3x1xf32>
+  vector.print %w : vector<3x1xf32>
+
+  return %result : vector<3x1xf32>
+}
+
+func.func @broadcast_low_dim_to_high_dim(%src: vector<3x1xf32>) -> vector<4x3x2xf32> {
+  // broadcast vector from low dimension to high can be transform into
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:51]: This is not completely clear. Can you please rephrase it.

File: examples/MLIRVector/vector-broadcast.mlir [hunk 6]
```diff
@@ -17,6 +19,117 @@ func.func @main() -> i32 {
...
+}
+
+func.func @broadcast_low_dim_to_high_dim(%src: vector<3x1xf32>) -> vector<4x3x2xf32> {
+  // broadcast vector from low dimension to high can be transform into 
+  // broadcasting two vectors with same dimension
+  %result = vector.broadcast %src : vector<3x1xf32> to vector<4x3x2xf32>
+
+  // vector<... x T> <=> vector<1 x ... x T> <=> vector<1 x 1 x ... x T>
+  // <=> vector<1 x 1 x ... x 1 x ... x T>
+  // it is obviously that we can extend a vector to any higher dimension by
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:57]: Typo: obviously -> obvious

File: examples/MLIRVector/vector-broadcast.mlir [hunk 7]
```diff
@@ -17,6 +19,117 @@ func.func @main() -> i32 {
...
+  // %result == %t2
+  vector.print %result : vector<4x3x2xf32>
+  vector.print %t2 : vector<4x3x2xf32>
+
+  return %result : vector<4x3x2xf32>
+}
+
+func.func @broadcast_1_to_n_case(%src: vector<1x3x1xf32>) -> vector<4x3x2xf32> {
+  // Case 1, "1->n" kind. 
+  //    broadcast %src: vector<1 x ... x T> to vector<n x ... x T>
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:75]: Please remove the unnecessary whitespace at the start of this comment. Take care of this at all relevant places in the PR.

File: examples/MLIRVector/vector-broadcast.mlir [hunk 8]
```diff
@@ -17,6 +19,117 @@ func.func @main() -> i32 {
...
+  %w0 = vector.insert %t0, %w_[0] : vector<2xf32> into vector<3x2xf32>
+  %w1 = vector.insert %t1, %w0[1] : vector<2xf32> into vector<3x2xf32>
+  %w2 = vector.insert %t2, %w1[2] : vector<2xf32> into vector<3x2xf32>
+
+  // now %w2 == %result
+  vector.print %result : vector<3x2xf32>
+  vector.print %w2 : vector<3x2xf32>
+
+  return %result : vector<3x2xf32>
+}
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:135]: Please add a newline here.

File: examples/MLIRVector/vector-type-cast.mlir [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+
+  // Normal usage
+  %m0 = vector.type_cast %mem0 : memref<2x4xi32> to memref<vector<2x4xi32>>
+  
+  %v0 = memref.load %m0[] : memref<vector<2x4xi32>>
+  vector.print %v0 : vector<2x4xi32>
+
+
+  // LLVM backend only support 1-D vectors and require they to align to A byte
```
Comment (meshtag) [inline examples/MLIRVector/vector-type-cast.mlir:35]: Typo: they -> them

File: examples/MLIRVector/vector-type-cast.mlir [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  // in an array, where A = length of vector rounded up to power of 2.  
+  // In current lowering path, MLIR's n-D vector<a... x b x T> is translated to
+  // LLVM's [a... x vector<b x T>]. 
+
+  // So if we want to tell compiler that "this memory region should be viewed as a 
+  // vector<a... x b x T>", we should manually align the vectors, which 
+  // unfortunately will break the abstraction that memref provided and introduce
+  // too much implementation details. 
+
+  // This is really buggy, so apply vector.type_cast to memref<a... x b x T>
```
Comment (meshtag) [inline examples/MLIRVector/vector-type-cast.mlir:45]: Typo: apply -> application of

File: examples/MLIRVector/vector-type-cast.mlir [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  // we init the memref and storing things into it.
+  %m2 = vector.type_cast %mem2 : memref<2xvector<3xi32>> to memref<vector<2x3xi32>>
+  
+  %v2 = memref.load %m2[] : memref<vector<2x3xi32>>
+  vector.print %v2 : vector<2x3xi32>
+
+
+  %ret = arith.constant 0 : i32
+  return %ret : i32
+}
```
Comment (meshtag) [inline examples/MLIRVector/vector-type-cast.mlir:67]: Please add a newline here.

**Review by meshtag — CHANGES_REQUESTED** [review]
Thanks for the PR! 
I have suggested some changes. Please take care of all typos, you can use tools like [Quillbot](https://quillbot.com/), to spend less time on these things and improve your productivity.

**Comment by Origami404** [comment]
Thank you for checking all my typos and grammar mistakes, and sorry for my poor English 🥲. To fix as more as them, I:

- Using VSCode extensions `cSpell` and `LTeX` to check the spelling and grammar mistakes inline
- Wrote a small script to extract comments from examples and merge them into paragraphs, then check them by QuillBot and Grammarly like normal articles
- Manually check all the examples again

Hopefully these approaches can save my poor English and minimize the number of typos!

I explain `vector.broadcast` by giving an example according to its recursive definition. A more "natural language-ed" way may sound like "first project itself into space with target rank, then repeat other part to fill target dimension length for each signle element dimension". But I think the recursive transformation way is clear.

I split all the typo fixes into another commit for the convenience of the reviewer. Once the review is done, I recommend squaring them into one commit for clearer commit history. (Three similar "typo fix" commits from three PRs seem redundant)

File: examples/MLIRVector/vector-type-cast.mlir (hunk 10 above)
Comment (Origami404) [reply examples/MLIRVector/vector-type-cast.mlir]: Changed to "applying"

File: examples/MLIRVector/vector-broadcast.mlir (hunk 7 above)
Comment (Origami404) [reply examples/MLIRVector/vector-broadcast.mlir:78]: I use these indentions as "code block" in Markdown for pseudo codes and variables that I want to emphasize. Is this an unrecommended usage?

File: examples/MLIRVector/vector-broadcast.mlir (hunk 4 above)
Comment (Origami404) [reply examples/MLIRVector/vector-broadcast.mlir]: Changed to `equals to splat it, like "std::fill" in C++.`.

File: examples/MLIRVector/vector-bitcast.mlir (hunk 3 above)
Comment (Origami404) [reply examples/MLIRVector/vector-bitcast.mlir:48]: I adjusted the description and merged them in the latest commit.

File: examples/MLIRVector/vector-broadcast.mlir (hunk 5 above)
Comment (Origami404) [reply examples/MLIRVector/vector-broadcast.mlir]: The sentence acts as a summary for this case, and the behavior it summarizes is described in detail below. Can you please give a little more detail about why you think it's unclear?

This sentence has had minor revisions to improve clarity in latest commit.

File: examples/MLIRVector/vector-type-cast.mlir (hunk 11 above)
Comment (meshtag) [reply examples/MLIRVector/vector-type-cast.mlir:67]: Gentle ping.

File: examples/MLIRVector/vector-broadcast.mlir [hunk 12]
```diff
@@ -8,15 +8,136 @@ func.func @main() -> i32 {
...
+  %t0 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  %t1 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  %t2 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  %t3 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  
+  // then create vector as [%t0, %t1, %t2, %t3]
+  %zero = arith.constant 0.0 : f32
+  %w_ = vector.broadcast %zero : f32 to vector<4x3x2xf32>
+  %w0 = vector.insert %t0, %w_[0] : vector<3x2xf32> into vector<4x3x2xf32>
+  %w1 = vector.insert %t0, %w0[1] : vector<3x2xf32> into vector<4x3x2xf32>
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:97]: Shouldn't this be `t1`? Below ones should be `t2` and `t3` right?

File: examples/MLIRVector/vector-broadcast.mlir [hunk 13]
```diff
@@ -8,15 +8,136 @@ func.func @main() -> i32 {
...
+
+  // %w == %result
+  vector.print %result : vector<3x1xf32>
+  vector.print %w : vector<3x1xf32>
+
+  return %result : vector<3x1xf32>
+}
+
+func.func @broadcast_low_dim_to_high_dim(%src: vector<3x1xf32>) -> vector<4x3x2xf32> {
+  // broadcasting vector wit smaller rank to larger one can be transformed into
```
Comment (meshtag) [inline examples/MLIRVector/vector-broadcast.mlir:54]: wit -> with

**Review by meshtag — APPROVED** [review]
LGTM, modulo minor changes :) 
Good effort!

File: examples/MLIRVector/vector-type-cast.mlir (hunk 11 above)
Comment (Origami404) [reply examples/MLIRVector/vector-type-cast.mlir]: fixed

File: examples/MLIRVector/vector-broadcast.mlir (hunk 12 above)
Comment (Origami404) [reply examples/MLIRVector/vector-broadcast.mlir]: My apologies, it should be `t1`, `t2`, ....
